### PR TITLE
Synthetics - initial revision

### PR DESCRIPTION
**Pull request by mateuszmidor**
KNTK-194

### CODE REVIEW

**Review by Fenthick — DISMISSED** [review]

**Comment by mmac-m3a** [comment]
Aren't most of the files in this pull request actually generated from OpenAPI/Swagger specs? If so, do they really need to be checked in? If so, why? We will have a public repo with API specs soon(ish). I strongly prefer all machine generated files (including code) to not be checked in and rather regenerated on every build. I see following benefits of this approach:
- it makes it clear what is the source of truth
- it helps to avoid the temptation to hand-tweak machine generated code (which is never a good idea)
- it makes code reviews sane (reviewing machine generated output is mostly pointless)
- it makes searches in the repo useful and faster

**Comment by Fenthick** [comment]
> Aren't most of the files in this pull request actually generated from OpenAPI/Swagger specs? If so, do they really need to be checked in? If so, why? We will have a public repo with API specs soon(ish). I strongly prefer all machine generated files (including code) to not be checked in and rather regenerated on every build. I see following benefits of this approach:
> 
>     * it makes it clear what is the source of truth
> 
>     * it helps to avoid the temptation to hand-tweak machine generated code (which is never a good idea)
> 
>     * it makes code reviews sane (reviewing machine generated output is mostly pointless)
> 
>     * it makes searches in the repo useful and faster

I think that in this case it is worth to check in generated code, because `apiv6/kentikapi` is a library that might be used by multiple projects/applications. Keeping generated code of implementation makes it usable instantly by calling `go get github.com/kentik/community_sdk_golang/apiv6` which, I think, is a huge deal. It makes the library ready-to-use.

If `apiv6/kentikapi` is going to be used by multiple projects/applications I think that keeping library ready-to-use outweighs cons that you listed.
If `apiv6/kentikapi` is going to be used only for TF providers I agree that "go gettability" is not big enough argument and we don't need to commit the generated code (we will manage to setup the build scripts in providers).

BTW, CI could check that re-generation of commited code produces null diff to verify the checked-in code. It would make hand-tweaking generated code fail the CI.

**Review by mmac-m3a — CHANGES_REQUESTED** [review]
Please rebuild the initial pull request based on the (now publicly accessible) [API spec](https://github.com/kentik/api-schema-public/blob/master/gen/openapiv2/kentik/synthetics/v202101beta1/synthetics.swagger.json). It is better for the initial version to be as clean as possible. Is there any way to avoid checking in the "non-code" generated files, like the documentation? I don't think that users will expect to find it in this repo (unlike the the generated golang code). I also prefer to not duplicaste the API specs (Swagger, OpenAPI) in this repo. Clearly, the code will be build based on them and we need to point interested parties to the source `api-schema-public` (say in the README).

**Comment by mateuszmidor** [comment]
PR Updated to comments.
@mmac-m3a can you please provide an example of valid Create Test request? I'm struggling getting Create Test to work. I keep receiving:
{"code":13,"message":"internal%20error%20(name:Error%20type:object)","details":[]}
when I do following request (payload taken from a test that was created beforehand in KentikPortal, changed the test name):

curl --location --request POST 'https://cloudexports.api.kentik.com/synthetics/v202101beta1/tests' \
--header 'X-CH-Auth-API-Token: <my token here>' \
--header 'X-CH-Auth-Email: mateusz.midor@codilime.com' \
--header 'Content-Type: application/json' \
--data-raw '{
    "test": {
        "name": "example-test-1",
        "type": "ip-address",
        "deviceId": "75702",
        "status": "TEST_STATUS_ACTIVE",
        "settings": {
            "ip": {
                "targets": [
                    "63.33.239.42"
                ]
            },
            "agentIds": [
                "658"
            ],
            "period": 0,
            "count": 0,
            "expiry": 0,
            "limit": 0,
            "tasks": [
                "ping",
                "traceroute"
            ],
            "healthSettings": {
                "latencyCritical": 0,
                "latencyWarning": 0,
                "packetLossCritical": 0,
                "packetLossWarning": 0,
                "jitterCritical": 0,
                "jitterWarning": 0,
                "httpLatencyCritical": 0,
                "httpLatencyWarning": 0,
                "httpValidCodes": [],
                "dnsValidCodes": []
            },
            "monitoringSettings": {
                "activationGracePeriod": "2",
                "activationTimeUnit": "m",
                "activationTimeWindow": "5",
                "activationTimes": "3",
                "notificationChannels": []
            },
            "ping": {
                "period": 60,
                "count": 5,
                "expiry": 3000
            },
            "trace": {
                "period": 60,
                "count": 3,
                "protocol": "udp",
                "port": 33434,
                "expiry": 22500,
                "limit": 30
            },
            "port": 443,
            "protocol": "icmp",
            "family": "IP_FAMILY_DUAL",
            "servers": [],
            "useLocalIp": false,
            "reciprocal": false,
            "rollupLevel": 1
        },
        "expiresOn": "2021-04-28T12:00:03.097Z",
        "cdate": "2021-04-16T12:00:02.084Z"
    }
}'

File: apiv6/README.md [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
+- <https://github.com/kentik/api-schema-public/tree/master/gen/openapiv2/kentik/synthetics/v202101beta1>
 
-Implemented API resources:
-- cloud export
-- synthetics
\ No newline at end of file
+## Generate client from specification
+
+Strictly saying, the code is generated from OpenAPI 3.0.0 specification.  
+To convert Swagger 2.0 -> OpenAPI 3.0.0, one can use the following converter: <https://mermade.org.uk/openapi-converter>
```
Comment (Fenthick) [inline apiv6/README.md:10]: Ideally we could have a script that downloads the latest openapi 2.0, converts to openapis 3.0 and generates the code/documents. What do you think about putting such task to backlog?

File: apiv6/generate_server.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    rm "$output_dir/go.mod"
+    rm "$output_dir/README.md"
+
+    echo "Done"
+}
+
+# GENERATE CLOUDEXPORT
+cloudexport_package="cloudexportstub"
+cloudexport_spec="cloud_export.openapi.yaml"
+cloudexport_server_output_dir="localhost_apiserver/cloud_export"
```
Comment (Fenthick) [inline apiv6/generate_server.sh:42]:
```suggestion
cloudexport_server_output_dir="stub_api_server/cloud_export"
```
This is subjective, but IMO the stub prefix is valuable information, and "localhost" is not.

File: apiv6/generate_server.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+# Generate golang server stub from OpenAPI 3.0.0 spec
```
Comment (Fenthick) [inline apiv6/generate_server.sh:3]: I see following files on the branch:
```bash
apiv6/localhost_apiserver
├── api
│   └── openapi.yaml
├── CloudExportStorage.json
├── go
│   ├── api_cloud_export_admin_service.go
│   ├── api_cloud_export_admin_service_service.go
│   ├── api.go
│   ├── helpers.go
│   ├── impl.go
│   ├── logger.go
│   ├── model_cloud_exportv202101beta1_status.go
│   ├── model_googlerpc_status.go
│   ├── model_protobuf_any.go
│   ├── model_v202101beta1_aws_properties.go
│   ├── model_v202101beta1_azure_properties.go
│   ├── model_v202101beta1_bgp_properties.go
│   ├── model_v202101beta1_cloud_export.go
│   ├── model_v202101beta1_cloud_export_type.go
│   ├── model_v202101beta1_create_cloud_export_request.go
│   ├── model_v202101beta1_create_cloud_export_response.go
│   ├── model_v202101beta1_gce_properties.go
│   ├── model_v202101beta1_get_cloud_export_response.go
│   ├── model_v202101beta1_ibm_properties.go
│   ├── model_v202101beta1_list_cloud_export_response.go
│   ├── model_v202101beta1_patch_cloud_export_request.go
│   ├── model_v202101beta1_patch_cloud_export_response.go
│   ├── model_v202101beta1_update_cloud_export_request.go
│   ├── model_v202101beta1_update_cloud_export_response.go
│   ├── repo.go
│   └── routers.go
├── main.go
├── README.md
├── run_server.sh
└── run_swagger_ui.sh
```

So we have generated files for cloudexport, but no for synthetics. As we are trying to minimize checked in generated files, could you delete those for cloudexport that are already checked in? Or are you planning to do that in separate PR?

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: apiv6/generate_server.sh (hunk 3 above)
Comment (Fenthick) [reply apiv6/generate_server.sh:3]: Ok, I see that this code is only partially generated so we need to check that in. Resolved.

File: apiv6/README.md (hunk 1 above)
Comment (mateuszmidor) [reply apiv6/README.md:10]: We can consider such an improvement if it feels worthwhile

File: apiv6/generate_server.sh (hunk 2 above)
Comment (mateuszmidor) [reply apiv6/generate_server.sh:42]: We can change the name, just remember to also  update cloudexport provider repo contents as it depends on this localhost_apiserver

File: apiv6/generate_server.sh (hunk 3 above)
Comment (mateuszmidor) [reply apiv6/generate_server.sh:3]: localhost_apiserver is required by cloudexport provider.
Support for synthetics will be added soon.

File: apiv6/README.md (hunk 1 above)
Comment (mmac-m3a) [reply apiv6/README.md:10]: I like the idea. Any automation that removes repeated mechanical work is helpful. I'll also discuss with engineering options for publishing OpenAPI 3.0.0 compatible spec to avoid the need for conversion.

**Review by mmac-m3a — COMMENTED** [review]
> PR Updated to comments.
> @mmac-m3a can you please provide an example of valid Create Test request? I'm struggling getting Create Test to work. I keep receiving:
> {"code":13,"message":"internal%20error%20(name:Error%20type:object)","details":[]}
> when I do following request (payload taken from a test that was created beforehand in KentikPortal, changed the test name):
> 
> curl --location --request POST 'https://cloudexports.api.kentik.com/synthetics/v202101beta1/tests'
> --header 'X-CH-Auth-API-Token: my token here'
> --header 'X-CH-Auth-Email: [mateusz.midor@codilime.com](mailto:mateusz.midor@codilime.com)'
> --header 'Content-Type: application/json'
...

Looks like you are executing the request against the wrong API end-point: `https://cloudexports.api.kentik.com`. It should be `https://synthetics.api.kentik.com/synthetics`. BTW, I also don't like the fact that the hostname is different for different APIs and I'm, in discussion with engineering to eventually change it.

File: apiv6/README.md [hunk 4]
```diff
@@ -1,7 +1,14 @@
...
 
-Implemented API resources:
-- cloud export
-- synthetics
\ No newline at end of file
+## Generate client from specification
+
+Strictly saying, the code is generated from OpenAPI 3.0.0 specification.  
+To convert Swagger 2.0 -> OpenAPI 3.0.0, one can use the following converter: <https://mermade.org.uk/openapi-converter>  
+One can use following shell scripts to generate client, server and markdown documentation:
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply apiv6/README.md:11]: Updating README

File: apiv6/README.md (hunk 1 above)
Comment (mateuszmidor) [reply apiv6/README.md:10]: OpenAPI 3.0.0 is the most recent and covers the Swagger spec functionality. Maybe it's a good idea to turn to OpenAPI 3.0.0 spec altogether. Just thinking

File: apiv6/generate_server.sh (hunk 2 above)
Comment (mateuszmidor) [reply apiv6/generate_server.sh:42]: The server stubs for synthetics and cloudexport are generated separately as they come from separate swagger api secs, but I will make localhost_apiserver support both cloudexport and synthetics, not sure if this is what you mean?

**Comment by mateuszmidor** [comment]
> > PR Updated to comments.
> > @mmac-m3a can you please provide an example of valid Create Test request? I'm struggling getting Create Test to work. I keep receiving:
> > {"code":13,"message":"internal%20error%20(name:Error%20type:object)","details":[]}
> > when I do following request (payload taken from a test that was created beforehand in KentikPortal, changed the test name):
> > curl --location --request POST 'https://cloudexports.api.kentik.com/synthetics/v202101beta1/tests'
> > --header 'X-CH-Auth-API-Token: my token here'
> > --header 'X-CH-Auth-Email: [mateusz.midor@codilime.com](mailto:mateusz.midor@codilime.com)'
> > --header 'Content-Type: application/json'
> > ...
> 
> Looks like you are executing the request against the wrong API end-point: `https://cloudexports.api.kentik.com`. It should be `https://synthetics.api.kentik.com/synthetics`. BTW, I also don't like the fact that the hostname is different for different APIs and I'm discussing with engineering to eventually change it.

Good point with the URL, I will update the code to accommodate for that.
Unfortunately, hitting https://synthetics.api.kentik.com/synthetics/v202101beta1/tests results in same error. I've already tried a number of different payloads describing the to create, but with no luck; an example of working test create request would come in handy here

File: apiv6/generate_server.sh (hunk 2 above)
Comment (mmac-m3a) [reply apiv6/generate_server.sh:42]: Yes, that is what I mean.

File: apiv6/localhost_apiserver/main.go [hunk 5]
```diff
@@ -1,38 +1,96 @@
...
+			"%s %s %s %s",
+			r.Method,
+			r.RequestURI,
+			name,
+			time.Since(start),
+		)
+	})
+}
+
+func makeRouter(cloudexporRepo *cloudexportstub.CloudExportRepo, syntheticsRepo *syntheticsstub.SyntheticsRepo) *mux.Router {
```
Comment (Fenthick) [inline apiv6/localhost_apiserver/main.go:31]: Please move the function below its usage.

Also:
```suggestion
func makeRouter(cr *cloudexportstub.CloudExportRepo, sr *syntheticsstub.SyntheticsRepo) *mux.Router {
```
to not repeat the name `CloudExportRepo` and decrease verbosity

File: apiv6/localhost_apiserver/main.go [hunk 6]
```diff
@@ -1,38 +1,96 @@
...
+	"github.com/gorilla/mux"
 
-	cloudexportstub "github.com/kentik/community_sdk_golang/apiv6/localhost_apiserver/go"
+	cloudexportstub "github.com/kentik/community_sdk_golang/apiv6/localhost_apiserver/cloudexport/go"
+	syntheticsstub "github.com/kentik/community_sdk_golang/apiv6/localhost_apiserver/synthetics/go"
 )
 
-func main() {
-	address, storageFilePath := readAddressStorage()
+func Logger(inner http.Handler, name string) http.Handler {
```
Comment (Fenthick) [inline apiv6/localhost_apiserver/main.go:15]: Please move the function below its usage and unexport it.

**Review by Fenthick — COMMENTED** [review]
Left stylistic comments to the new code

File: apiv6/localhost_apiserver/main.go (hunk 5 above)
Comment (Fenthick) [reply apiv6/localhost_apiserver/main.go]: Resolved

File: apiv6/localhost_apiserver/main.go (hunk 6 above)
Comment (Fenthick) [reply apiv6/localhost_apiserver/main.go]: Resolved

**Review by Fenthick — APPROVED** [review]

**Review by mmac-m3a — COMMENTED** [review]
Looks like generated content, specifically API documentation crept back into the PR. Now we even have 2 sets for synthetics.
It also looks like the fixes to the Swagger spec have been lost. For example, PatchTest now has valid body definition.

I'm holding up this pull request because:
- I don't want to pollute the public repo with (not yet) functional code (e.g. CreateTest), due to problems on Kentik side (which, conceivably, may need changes to the Swagger spec),
- I don't want the API documentation to be duplicated in the SDK repo
- we need to seriously consider relying on the pre-generated code in the api-schema-public repo, which make require significant changes to the code (I assume)

File: apiv6/examples/synthetics/main.go [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+	"os"
+	"time"
+
+	"github.com/kentik/community_sdk_golang/apiv6/examples"
+	"github.com/kentik/community_sdk_golang/apiv6/kentikapi"
+	"github.com/kentik/community_sdk_golang/apiv6/kentikapi/synthetics"
+)
+
+// Current status:
+// CreateAgent                     POST   - missing request payload data struct in swagger spec
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply apiv6/examples/synthetics/main.go]: Hmm, It might be the case that you are looking at some older revision of the code, the recent revision is Apr 22, 2021 "Added support for synthetics in localhost_apiserver" (I did a correcting overriding git push there, seems this introduced confusion; will avoid doing this again). Current status:
- generated docs removed
- spec files removed
- recent spec from public repo applied (PATCH works)
- examples for synthetics work (except for CREATE tests)

File: apiv6/examples/synthetics/main.go [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+// CreateTest                      POST   - api call returns http 500
+// DeleteAgent                     DELETE - OK
+// DeleteTest                      DELETE - OK
+// GetAgent                        GET    - OK
+// GetTest                         GET    - OK
+// ListAgents                      GET    - OK
+// ListTests                       GET    - OK
+// PatchAgent                      PATCH  - api call returns  http 500
+// PatchTest                       PATCH  - missing payload for the request in swagger spec
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply apiv6/examples/synthetics/main.go]: Updated.

File: apiv6/examples/synthetics/main.go [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+	createReq := client.SyntheticsAdminServiceApi.TestCreate(context.Background()).V202101beta1CreateTestRequest(createReqPayload)
+	createResp, httpResp, err := createReq.Execute()
+	if err != nil {
+		return fmt.Errorf("%v %v", err, httpResp)
+	}
+	examples.PrettyPrint(createResp)
+	fmt.Println()
+	created := *createResp.Test
+
+	// NOTE: PATCH is missing the request payload in swagger spec, so currently doesnt compile.
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply apiv6/examples/synthetics/main.go]: Updated.

File: apiv6/examples/synthetics/main.go [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+	if err != nil {
+		return fmt.Errorf("%v %v", err, httpResp)
+	}
+	examples.PrettyPrint(getResp)
+	fmt.Println()
+
+	return nil
+}
+
+// NOTE: PATCH is missing the request payload in swagger spec, so currently doesnt compile
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply apiv6/examples/synthetics/main.go]: Updated.

File: kentikapi/internal/api_payloads/device.go [hunk 11]
```diff
@@ -177,19 +177,19 @@ func deviceBGPTypeFromStringPtr(s *string) *models.DeviceBGPType {
...
-		SendingIPS:          d.SendingIPS,
-		PlanID:              d.PlanID,
-		SiteID:              d.SiteID,
-		CDNAttr:             cdnAttributeToStringPtr(d.CDNAttr),
-		DeviceDescription:   d.DeviceDescription,
-		DeviceSNMNPIP:       d.DeviceSNMNPIP,
-		DeviceSNMPCommunity: d.DeviceSNMPCommunity,
-		DeviceSNMPv3Conf:    snmp3ConfToPayload(d.DeviceSNMPv3Conf),
-		MinimizeSNMP:        d.MinimizeSNMP,
+		DeviceName:            &d.DeviceName,
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply kentikapi/internal/api_payloads/device.go:180]: Good idea

**Comment by mateuszmidor** [comment]
Synthetics functionality delivered in PR 35
